alchemiscale/storage/statestore.py

```diff
@@ -242,138 +242,139 @@ def reset(self):
+                node_b,
+                _org=scope.org,
+                _campaign=scope.campaign,
+                _project=scope.project,
+                **kwargs,
+            )
+            relationships.append(rel)
+
+        def handle_list(node, key, values):
+            if isinstance(values[0], (int, float, str)) and all(
+                (isinstance(x, type(values[0])) for x in values)
+            ):
+                node[key] = values
+                return
+            # list of gufe key: [{":gufe-key:": ...}, {":gufe-key:": ...}, {":gufe-key:": ...}]
+            elif all(map(is_gufe_dict, values)):
+                for i, x in enumerate(values):
+                    rel_key = x[":gufe-key:"]
+                    update_relationships(
+                        node, get_previous_node(rel_key), attribute=key, index=i
                     )
-                    | subgraph_
-                )
+            else:
+                node[key] = json.dumps(values, cls=JSON_HANDLER.encoder)
+                node["_json_props"].append(key)
 
+        def handle_dict(node, key, dct):
+            # e.g. {'ligand': {':gufe-key:': 'SmallMoleculeComponent-abc123'}}
+            if all(map(is_gufe_dict, dct.values())):
+                for k, v in dct.items():
+                    rel_key = v[":gufe-key:"]
+                    update_relationships(
+                        node, get_previous_node(rel_key), attribute=key, key=k
+                    )
             else:
-                node[key] = value
+                node[key] = json.dumps(dct, cls=JSON_HANDLER.encoder)
+                node["_json_props"].append(key)
 
-        subgraph = subgraph | node
+        def handle_tuple(node, key, values):
+            if not (
+                isinstance(values[0], (int, float, str))
+                and all((isinstance(x, type(values[0])) for x in values))
+            ):
+                node[key] = json.dumps(values, cls=JSON_HANDLER.encoder)
+                node["_json_props"].append(key)
+
+        def process_keyed_dict(gufe_key, kd):
+            node = Node("GufeTokenizable", kd["__qualname__"])
+            node["_json_props"] = []
+            for key, value in kd.items():
+                match value:
+                    case {":gufe-key:": rel_key}:
+                        update_relationships(
+                            node, get_previous_node(rel_key), attribute=key
+                        )
+                    case dict():
+                        handle_dict(node, key, value)
+                    case list():
+                        handle_list(node, key, value)
+                    case tuple():
+                        handle_tuple(node, key, value)
+                    case SettingsBaseModel():
+                        node[key] = json.dumps(
+                            value, cls=JSON_HANDLER.encoder, sort_keys=True
+                        )
+                        node["_json_props"].append(key)
```

Comment:
Perhaps put this in its own function for consistency with the others?

Reply:
sure